File: 2022/10/solve.py
from pathlib import Path


class CustomStopError(Exception):
    pass


def solve_one(data: str) -> int:
    CYCLES = (20, 60, 100, 140, 180, 220)

    x = 1
    cycle = 0
    total = 0

    def next_cycle(count: int | None = None) -> None:
        nonlocal x, cycle, total
        cycle += 1

        if cycle in CYCLES:
            total += x * cycle

        if count is not None:
            x += count

        if cycle == CYCLES[-1]:
            raise CustomStopError

    for command in data.split("\n"):
        command = command.rstrip("\n")
        try:
            if command == "noop":
                next_cycle()
            else:
                count = int(command.split(" ")[-1])
                next_cycle()
                next_cycle(count)
        except CustomStopError:
            return total

    raise AssertionError("This should never happen")


def solve_two(data: str) -> None:
    x = 1
    cycle = 0
    pixel = 0

    def next_cycle(count: int | None = None) -> None:
        nonlocal x, cycle, pixel
        cycle += 1

        if x - 1 <= pixel <= x + 1:
            print("#", end="")
        else:
            print(".", end="")

        pixel += 1

        if count is not None:
            x += count

        if cycle > 0 and cycle % 40 == 0:
            print()
            pixel = 0

        if cycle == 240:
            raise CustomStopError

    for command in data.split("\n"):
        command = command.rstrip("\n")
        try:
            if command == "noop":
                next_cycle()
            else:
                count = int(command.split(" ")[1])
                next_cycle()
                next_cycle(count)
        except CustomStopError:
            pass


if __name__ == "__main__":
    data = (Path(__file__).parent / "input.txt").read_text().strip()

    solution_one = solve_one(data)
    print(solution_one)
    assert solution_one == 15140

    solution_two = solve_two(data)
    # print(solution_two)
    # assert solution_two == ...

File: 2021/01/solve.swift
import Foundation

let data = try String(contentsOfFile: "input.txt")

let result_one = solve_one(data: data)
print(result_one)
assert(result_one == 1288)

let result_two = solve_two(data: data)
print(result_two)
assert(result_two == 1311)

func solve_one(data: String) -> Int {
    solve(data: data, count: 1)
}

func solve_two(data: String) -> Int {
    solve(data: data, count: 3)
}

func solve(data: String, count: Int) -> Int {
    let measurements = parse_input(data: data)
    var increases = 0

    for i in count..<measurements.count {
        if measurements[i] > measurements[i - count] {
            increases += 1
        }
    }

    return increases
}

func parse_input(data: String) -> [Int] {
    data
        .split(separator: "\n")
        .map { Int($0)! }
}
